File: backend/app/schemas/pipeline.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field

from app.schemas.social_account import PublishDraftResponse, SocialAccountResponse


class RemixConfigRequest(BaseModel):
    target_duration_seconds: Optional[float] = Field(default=None, ge=1, le=300)
    mood: Optional[str] = Field(default=None, max_length=80)
    bgm_material_id: Optional[str] = None
    bgm_mood: Literal["none", "upbeat", "calm", "cinematic", "energetic"] = "cinematic"
    bgm_volume: float = Field(default=0.15, ge=0, le=1)
    include_source_audio: bool = False
    add_voiceover: bool = True
    voiceover_script: Optional[str] = Field(default=None, max_length=8000)


class PipelineCreateRequest(BaseModel):
    script: str = Field(default="", max_length=8000)  # free-form user input: script or requirement description
    image_ids: list[str] = Field(default_factory=list, max_length=100)  # Material IDs
    session_id: Optional[str] = None
    background_template_id: Optional[str] = None
    reference_video_id: Optional[str] = None  # VideoUpload ID for replication mode
    reference_video_ids: list[str] = Field(default_factory=list, max_length=20)  # VideoUpload IDs for remix mode
    remix_config: Optional[RemixConfigRequest] = None
    platform: Literal["douyin", "xiaohongshu", "bilibili", "generic"] = "generic"
    duration_seconds: int = Field(default=30, ge=1, le=300)
    duration_mode: Literal["auto", "fixed"] = "fixed"
    no_audio: bool = True  # Legacy audio flag used when explicit audio flags are omitted.
    video_model_no_audio: Optional[bool] = None  # True = Seedance/Kling generates silent clips.
    voiceover_no_audio: Optional[bool] = None  # True = skip VidGen TTS/subtitle generation.
    generation_model: Literal["seedance1.5-pro", "seedance2.0", "kling", "mock"] = "seedance1.5-pro"
    style: Literal["commercial", "lifestyle", "cinematic"] = "commercial"
    voice_id: str = Field(default="default", min_length=1, max_length=80)
    transition: Literal["none", "fade", "dissolve", "slideright", "slideup"] = "none"
    transition_duration: float = Field(default=0.5, ge=0, le=5)
    bgm_mood: Literal["none", "upbeat", "calm", "cinematic", "energetic"] = "none"
    bgm_volume: float = Field(default=0.15, ge=0, le=1)
    watermark_image_id: Optional[str] = None  # material ID of watermark/logo image
    review_prompts: Optional[bool] = None  # override human-in-the-loop prompt review for this run


class ConfirmPlanRequest(BaseModel):
    approved: bool
    adjustments: Optional[str] = Field(default=None, max_length=4000)  # user feedback when not approved


class RemixSegmentEdit(BaseModel):
    segment_idx: int = Field(ge=0)
    source_video_id: Optional[str] = None
    start_seconds: Optional[float] = Field(default=None, ge=0)
    end_seconds: Optional[float] = Field(default=None, ge=0)
    transition_type: Optional[Literal["cut", "fade", "dissolve", "slideright", "slideup"]] = None
    removed: bool = False


class ConfirmRemixRequest(BaseModel):
    approved: bool
    adjustments: Optional[str] = Field(default=None, max_length=4000)
    edited_segments: list[RemixSegmentEdit] = Field(default_factory=list, max_length=100)


class ShotEditRequest(BaseModel):
    shot_idx: int = Field(ge=0)
    script_segment: Optional[str] = Field(default=None, max_length=2000)
    video_prompt: Optional[str] = Field(default=None, max_length=4000)
    duration_seconds: Optional[float] = Field(default=None, ge=0.5, le=60)


class ConfirmPromptReviewRequest(BaseModel):
    edited_shots: Optional[list[ShotEditRequest]] = None


class RetryShotRequest(BaseModel):
    shot_indices: list[int] = Field(min_length=1, max_length=100)


class EstimateCostRequest(BaseModel):
    shot_plan: list[dict] = Field(default_factory=list, max_length=100)      # shot_prompts list
    model: Literal["seedance1.5-pro", "seedance2.0", "kling", "mock"] = "seedance1.5-pro"
    voiceover_no_audio: bool = False
    tts_char_count: int = Field(default=0, ge=0, le=20000)         # if provided, overrides shot_plan script length estimate
    bgm_mood: Literal["none", "upbeat", "calm", "cinematic", "energetic"] = "none"


class EstimateCostResponse(BaseModel):
    estimated_total_cny: float
    breakdown: dict  # {"video_gen": float, "tts": float, "llm": float, "bgm": float}
    shot_count: int
    total_generation_seconds: float
    warning: Optional[str] = None


class ScriptGenerateRequest(BaseModel):
    image_ids: list[str] = Field(min_length=1, max_length=100)


class ScriptGenerateResponse(BaseModel):
    script: str


class PrefightCheckRequest(BaseModel):
    script: str = Field(max_length=8000)
    image_count: int = Field(ge=0, le=100)
    duration_seconds: int = Field(default=30, ge=1, le=300)
    duration_mode: Literal["auto", "fixed"] = "fixed"


class PrefightCheckResponse(BaseModel):
    ok: bool
    warning: Optional[str] = None
    estimated_audio_seconds: float = 0
    max_video_seconds: int = 0
    recommended_image_count: int = 0
    estimated_tokens: int = 0  # rough total token estimate for the pipeline


class PipelineRunResponse(BaseModel):
    id: str
    project_id: str
    session_id: Optional[str] = None
    trace_id: str
    engine: str = "pipeline"
    status: str
    current_agent: Optional[str] = None
    overall_score: Optional[float] = None
    final_video_path: Optional[str] = None
    error_message: Optional[str] = None
    retry_count: int = 0
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class AgentExecutionResponse(BaseModel):
    id: str
    agent_name: str
    status: str
    attempt_number: int = 1
    input_data: Optional[dict] = None
    output_data: Optional[dict] = None
    duration_ms: Optional[int] = None
    error_message: Optional[str] = None
    progress_text: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class UsageByAgentResponse(BaseModel):
    agent_name: str
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    request_count: int = 0


class UsageByModelResponse(BaseModel):
    provider: str
    model_name: str
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    request_count: int = 0


class PipelineUsageResponse(BaseModel):
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    request_count: int = 0
    by_agent: list[UsageByAgentResponse] = Field(default_factory=list)
    by_model: list[UsageByModelResponse] = Field(default_factory=list)


class DeliveryActionRequest(BaseModel):
    social_account_id: Optional[str] = None
    title: Optional[str] = Field(default=None, max_length=80)
    description: Optional[str] = Field(default=None, max_length=2000)
    hashtags: list[str] = Field(default_factory=list, max_length=20)
    visibility: Literal["public", "private", "friends"] = "public"
    cover_title: Optional[str] = Field(default=None, max_length=80)


class PlatformPreviewCardResponse(BaseModel):
    platform: str
    label: str
    aspect_ratio: str
    recommended_resolution: str
    cover_title: str
    headline: str
    caption: str
    layout_hint: str
    safe_zone_tip: str
    context_hint: str
    primary_action: str


class VideoDeliveryResponse(BaseModel):
    id: str
    user_id: str
    project_id: str
    pipeline_run_id: str
    action_type: str
    platform: str
    status: str
    social_account_id: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    draft_payload: Optional[dict] = None
    saved_video_path: Optional[str] = None
    external_id: Optional[str] = None
    external_url: Optional[str] = None
    external_status: Optional[str] = None
    response_payload: Optional[dict] = None
    platform_error_code: Optional[str] = None
    error_message: Optional[str] = None
    submitted_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class PipelineDeliveryResponse(BaseModel):
    previews: list[PlatformPreviewCardResponse] = Field(default_factory=list)
    records: list[VideoDeliveryResponse] = Field(default_factory=list)
    connected_social_accounts: list[SocialAccountResponse] = Field(default_factory=list)
    recommended_publish_account: Optional[SocialAccountResponse] = None
    latest_publish_draft: Optional[PublishDraftResponse] = None


class PipelineArtifactResponse(BaseModel):
    id: str
    user_id: str
    project_id: str
    project_name: Optional[str] = None
    pipeline_run_id: str
    asset_key: str
    asset_type: str
    source_agent: str
    title: Optional[str] = None
    description: Optional[str] = None
    mime_type: Optional[str] = None
    file_path: Optional[str] = None
    file_url: Optional[str] = None
    file_size: Optional[int] = None
    text_content: Optional[str] = None
    metadata: dict = Field(default_factory=dict)
    duration_ms: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class ProjectPipelineUsageItem(BaseModel):
    id: str
    status: str
    current_agent: Optional[str] = None
    total_tokens: int = 0
    request_count: int = 0
    created_at: datetime


class ProjectUsageSummaryResponse(BaseModel):
    project_id: str
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    request_count: int = 0
    latest_pipeline_status: Optional[str] = None
    latest_current_agent: Optional[str] = None
    pipelines: list[ProjectPipelineUsageItem] = Field(default_factory=list)


class ArtifactFileResponse(BaseModel):
    name: str
    path: str
    url: str
    content: Optional[str] = None
    shot_idx: Optional[int] = None
    duration_ms: Optional[int] = None
    kind: Optional[str] = None


class PromptHistoryItemResponse(BaseModel):
    shot_idx: int
    script_segment: Optional[str] = None
    video_prompt: str
    duration_seconds: Optional[float] = None


class ProjectHistoryRunResponse(BaseModel):
    run_id: str
    status: str
    created_at: datetime
    completed_at: Optional[datetime] = None
    current_agent: Optional[str] = None
    total_tokens: int = 0
    request_count: int = 0
    input_script: Optional[str] = None
    voice_params: Optional[dict] = None
    prompts: list[PromptHistoryItemResponse] = []
    audio_files: list[ArtifactFileResponse] = []
    subtitle_files: list[ArtifactFileResponse] = []
    generated_videos: list[ArtifactFileResponse] = []
    final_videos: list[ArtifactFileResponse] = []


class ProjectHistoryResponse(BaseModel):
    project_id: str
    runs: list[ProjectHistoryRunResponse] = []

File: frontend/src/types/pipeline.ts
import type { MaterialItem, MaterialSelection, VideoUpload } from './assets'
import type { PublishDraft, SocialAccount } from './social'

export interface PipelineConfig {
  script: string
  image_ids: string[]
  session_id?: string | null
  reference_video_id?: string | null
  reference_video_ids?: string[]
  remix_config?: RemixConfig | null
  background_template_id?: string | null
  platform: string
  duration_seconds: number
  duration_mode?: string
  no_audio?: boolean
  video_model_no_audio?: boolean
  voiceover_no_audio?: boolean
  generation_model?: string
  style: string
  voice_id: string
  transition?: string
  transition_duration?: number
  bgm_mood?: string
  bgm_volume?: number
  watermark_image_id?: string | null
}

export interface RemixConfig {
  target_duration_seconds?: number | null
  mood?: string | null
  bgm_material_id?: string | null
  bgm_mood?: string
  bgm_volume?: number
  include_source_audio?: boolean
  add_voiceover?: boolean
  voiceover_script?: string | null
}

export interface RemixPlanSegment {
  segment_idx: number
  source_video_id: string
  source_shot_idx?: number
  start_seconds: number
  end_seconds: number
  description?: string
  script_segment?: string
  narration?: string
  voiceover?: string
  role?: string
  quality_score?: number
  transition_to_next?: string
  transition_duration?: number
  reference_keyframe_path?: string
  removed?: boolean
}

export interface RemixPlan {
  title?: string
  concept?: string
  target_duration_seconds?: number
  source_videos?: Array<Record<string, unknown>>
  segments: RemixPlanSegment[]
  audio_design?: Record<string, unknown>
  analysis_report?: string
}

export interface RemixSegmentEdit {
  segment_idx: number
  source_video_id?: string
  start_seconds?: number
  end_seconds?: number
  transition_type?: string
  removed?: boolean
}

export interface GenerateScriptResponse {
  script: string
}

export interface PipelineRun {
  id: string
  project_id: string
  session_id?: string | null
  trace_id: string
  engine: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'waiting_confirmation' | 'waiting_prompt_review' | 'waiting_remix_confirmation'
  current_agent: string | null
  overall_score: number | null
  final_video_path: string | null
  error_message: string | null
  retry_count: number
  created_at: string
  updated_at: string
  completed_at: string | null
}

export interface AgentExecution {
  id: string
  agent_name: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped' | 'cancelled'
  attempt_number: number
  input_data?: Record<string, unknown> | null
  output_data?: Record<string, unknown> | null
  duration_ms: number | null
  error_message: string | null
  progress_text?: string | null
  created_at: string
  completed_at: string | null
}

export interface PipelineUsageByAgent {
  agent_name: string
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  request_count: number
}

export interface PipelineUsageByModel {
  provider: string
  model_name: string
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  request_count: number
}

export interface PipelineUsageSummary {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  request_count: number
  by_agent: PipelineUsageByAgent[]
  by_model: PipelineUsageByModel[]
}

export interface AutoChatMessageImagePayload {
  id: string
  url: string
  name: string
}

export interface AutoChatMessageVideoPayload {
  id: string
  name: string
  streamUrl: string
}

export interface AutoChatMessageFilePayload {
  id: string
  name: string
  url: string
  mimeType?: string | null
}

export interface DirectorPlanShot {
  shot_idx: number
  duration_seconds: number
  duration_range_label?: string
  generation_duration_seconds?: number
  sequence_role?: string
  sequence_reason?: string
  shot_purpose?: string
  script_segment?: string
  video_prompt?: string
  source_image_idx?: number
}

export interface DirectorPlanVoiceDesign {
  voice_id?: string
  speed?: number
  tone?: string
}

export interface DirectorPlan {
  run_id: string
  shot_prompts: DirectorPlanShot[]
  voice_design: DirectorPlanVoiceDesign
  director_summary?: string
  creative_concept?: string
  pacing_strategy?: string
  narration_script?: string
}

export interface AutoChatMessagePayload {
  mutedLines?: string[]
  images?: AutoChatMessageImagePayload[]
  files?: AutoChatMessageFilePayload[]
  video?: AutoChatMessageVideoPayload | null
  publishDraft?: PublishDraft | null
  directorPlan?: DirectorPlan | null
}

export interface AutoChatSessionState {
  draft_script: string | null
  background_template_id: string | null
  reference_video_id: string | null
  reference_video_ids: string[]
  video_platform: string
  video_no_audio: boolean
  video_model_no_audio?: boolean
  voiceover_no_audio?: boolean
  generation_model?: string
  duration_mode: string
  video_transition: string
  bgm_mood: string
  watermark_id: string | null
  current_run_id: string | null
}

export interface AutoChatSessionSummary {
  id: string
  project_id: string
  title: string
  status_preview: string
  latest_message_excerpt: string | null
  latest_message_role: string | null
  reference_video_name: string | null
  current_run_id: string | null
  current_run_status: string | null
  last_activity_at: string
  created_at: string
  updated_at: string
}

export interface AutoChatSessionMessage {
  id: string
  session_id: string
  role: 'user' | 'assistant' | 'system'
  title?: string | null
  content: string
  payload?: AutoChatMessagePayload | null
  created_at: string
  updated_at: string
}

export interface AutoChatSessionDetail {
  session: AutoChatSessionSummary
  state: AutoChatSessionState
  messages: AutoChatSessionMessage[]
  selected_materials: MaterialSelection[]
  selected_material_items: MaterialItem[]
  reference_video: VideoUpload | null
  reference_videos: VideoUpload[]
  current_run: PipelineRun | null
  agent_executions: AgentExecution[]
  delivery_info: PipelineDeliveryInfo | null
  usage_summary: PipelineUsageSummary | null
  connected_social_accounts?: SocialAccount[]
  recommended_publish_account?: SocialAccount | null
  latest_publish_draft?: PublishDraft | null
}

export interface ProjectPipelineUsageItem {
  id: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  current_agent: string | null
  total_tokens: number
  request_count: number
  created_at: string
}

export interface ProjectUsageSummary {
  project_id: string
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  request_count: number
  latest_pipeline_status: string | null
  latest_current_agent: string | null
  pipelines: ProjectPipelineUsageItem[]
}

export interface ProjectArtifactFile {
  name: string
  path: string
  url: string
  content: string | null
  shot_idx: number | null
  duration_ms: number | null
  kind: string | null
}

export interface PromptHistoryItem {
  shot_idx: number
  script_segment: string | null
  video_prompt: string
  duration_seconds: number | null
}

export interface ProjectHistoryRun {
  run_id: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  created_at: string
  completed_at: string | null
  current_agent: string | null
  total_tokens: number
  request_count: number
  input_script: string | null
  voice_params: Record<string, unknown> | null
  prompts: PromptHistoryItem[]
  audio_files: ProjectArtifactFile[]
  subtitle_files: ProjectArtifactFile[]
  generated_videos: ProjectArtifactFile[]
  final_videos: ProjectArtifactFile[]
}

export interface ProjectHistoryResponse {
  project_id: string
  runs: ProjectHistoryRun[]
}

export interface PipelineDeliveryInfo {
  previews: PlatformPreviewCard[]
  records: VideoDeliveryRecord[]
  connected_social_accounts?: SocialAccount[]
  recommended_publish_account?: SocialAccount | null
  latest_publish_draft?: PublishDraft | null
}

export interface PlatformPreviewCard {
  platform: 'douyin' | 'youtube'
  label: string
  aspect_ratio: string
  recommended_resolution: string
  cover_title: string
  headline: string
  caption: string
  layout_hint: string
  safe_zone_tip: string
  context_hint: string
  primary_action: string
}

export interface VideoDeliveryRecord {
  id: string
  user_id: string
  project_id: string
  pipeline_run_id: string
  action_type: 'save' | 'publish'
  platform: string
  status: 'pending' | 'draft' | 'saved' | 'submitted' | 'published' | 'failed'
  social_account_id?: string | null
  title: string | null
  description: string | null
  draft_payload?: Record<string, unknown> | null
  saved_video_path: string | null
  external_id: string | null
  external_url: string | null
  external_status?: string | null
  response_payload?: Record<string, unknown> | null
  platform_error_code?: string | null
  error_message: string | null
  submitted_at?: string | null
  published_at?: string | null
  created_at: string
  updated_at: string
}
